File: 3rd Year/Software Engineering/Assignment 1/LCA.java
/*
    a
   / \
  b   c
 / \
d   e
   / \
  f   g
*/

public class LCA {
    public static void main(String[] args) {
        Node a = new Node('a', null);
        Node b = new Node('b', a);
        Node c = new Node('c', a);
        Node d = new Node('d', b);
        Node e = new Node('e', b);
        Node f = new Node('f', e);
        Node g = new Node('g', e);
        System.out.println("Should be 'b' - " + getLCA(d,f));
        System.out.println("Should be 'e' - " + getLCA(f,g));
        System.out.println("Should be 'a' - " + getLCA(g, c));
    }

    public static char getLCA(Node one, Node two) {
        if (one.hasParent() && two.hasParent()) {
            if (one.getParentKey() == two.getParentKey()) {
                return one.getParentKey();
            } else if (one.getTeir()<two.getTeir()) {
                return getLCA(one, two.getParent());
            } else {
                return getLCA(two, one.getParent());
            }
        } else {
            return '0';
        }
    }
}

File: 3rd Year/Software Engineering/Assignment 1/LCA.py
class Node:
    def __init__(self, key, parent):
        self.key = key
        self.parent = parent
        if parent is None:
            self.tier = 0;
        else:
            self.tier = parent.tier+1

a = Node('a', None)
b = Node('b', a)
c = Node('c', a)
d = Node('d', b)
e = Node('e', b)
f = Node('f', e)
g = Node('g', e)
h = Node('h', c)

def getLCA(one, two):
    if one.parent!=None and two.parent!=None:
        if one.parent.key==two.parent.key:
            return one.parent.key
        elif one.tier < two.tier:
            return getLCA(one, two.parent)
        else:
            return getLCA(two, one.parent)
    return 'none'

print('Should be b - ' + getLCA(d,f))
print('Should be e - ' + getLCA(f,g))
print('Should be a - ' + getLCA(g,c))
print('Should be a - ' + getLCA(d,h))
print('Should be none - ' + getLCA(a,b))